File: 338.counting-bits.py
#
# @lc app=leetcode id=338 lang=python3
#
# [338] Counting Bits
#

# @lc code=start
class Solution:
    def countBits(self, num: int) -> List[int]:
        res = [0]
        while len(res) <= num:
            res.extend([i+1 for i in res])
        return res[0:num+1]
# @lc code=end

File: 338.counting-bits.go
/*
 * @lc app=leetcode id=338 lang=golang
 *
 * [338] Counting Bits
 */
package main

import "fmt"

// @lc code=start
func countBits(n int) []int {
	hightBit := 0
	ret := []int{0}
	for i := 1; i <= n; i++ {
		if i&(i-1) == 0 {
			hightBit = i
			ret = append(ret, 1)
		} else {
			ret = append(ret, ret[i-hightBit]+1)
		}
	}
	return ret
}

// @lc code=end

func main() {
	fmt.Println(countBits(5))
}
